File: frontend/forge/src/types/report.ts
import { DataSource } from '@/types/common';
import type { Data as PlotlyData, Layout as PlotlyLayout } from 'plotly.js';

export type ReportFormat = 'json' | 'pdf' | 'html';

export interface DataQualityStats {
  unique_count: number;
  duplicate_percentage: number;
}

export interface DataQuality {
  completeness: {
    score: number;
    missing_values: Record<string, number>;
    total_records: number;
  };
  accuracy: {
    score: number;
    invalid_values: Record<string, number>;
  };
  consistency: {
    score: number;
    inconsistencies: Record<string, string[]>;
  };
}

export interface TimeSeriesAnalysis {
  trend: string;
  seasonality: string;
  outliers: number[];
  forecast?: number[];
  trends?: Record<string, {
    mean: number;
    std: number;
    trend: string;
    trend_strength: number;
    trend_components?: {
      trend: number[];
      seasonal: number[];
      residual: number[];
    };
  }>;
}

export interface NumericalAnalysis {
  mean: number;
  median: number;
  std: number;
  min: number;
  max: number;
  quartiles: [number, number, number];
}

export interface CategoricalAnalysis {
  unique_values: number;
  most_common: [string, number][];
  distribution: Record<string, number>;
}

export interface TrendAnalysis {
  direction: 'increasing' | 'decreasing' | 'stable';
  strength: number;
  confidence: number;
}

export interface NumericalStatistics {
  correlation_matrix?: Record<string, Record<string, number>>;
  summary_stats: Record<string, NumericalAnalysis>;
}

export interface StatisticalSummary {
  mean?: number;
  median?: number;
  std?: number;
  min?: number;
  max?: number;
  count?: number;
  percentiles?: Record<string, number>;
}

export interface AnalysisResult {
  llm_analysis: string;
  insights: {
    summary_stats: Record<string, StatisticalSummary>;
  };
  statistical_analysis?: {
    time_series?: TimeSeriesAnalysis;
    numerical?: NumericalAnalysis;
    categorical?: CategoricalAnalysis;
  };
  data_quality: DataQuality;
}

export interface VisualizationResponse {
  type: string;
  data: Array<Partial<PlotlyData>>;
  layout: Partial<PlotlyLayout>;
  config?: {
    responsive?: boolean;
    displayModeBar?: boolean;
    [key: string]: any;
  };
}

export interface StatisticalAnalysis {
  time_series: TimeSeriesAnalysis | null;
  numerical: NumericalAnalysis | null;
  categorical: CategoricalAnalysis | null;
}

export interface AnalysisInsights {
  summary_stats: Record<string, StatisticalSummary>;
  trends?: Record<string, TrendAnalysis>;
  correlations?: Record<string, number>;
}

export interface EnhancedReport {
  id: string;
  data_type: 'time_series' | 'numerical' | 'categorical' | 'mixed';
  analysis: {
    llm_analysis: string;
    insights: AnalysisInsights;
    statistical_analysis: StatisticalAnalysis;
    data_quality: DataQuality;
  };
  visualizations: VisualizationResponse[];
  metadata: {
    created_at: string;
    updated_at: string;
    source: {
      type: string;
      name: string;
    };
    rows: number;
    columns: number;
  };
}

export interface ReportGenerationOptions {
  format: 'json' | 'pdf' | 'html';
  includeVisualizations: boolean;
  includeTables: boolean;
  customizations?: {
    theme?: 'light' | 'dark';
    fontSize?: number;
    colors?: string[];
  };
}

export interface DataProcessingOptions {
  cleanMissingValues?: boolean;
  removeOutliers?: boolean;
  normalizeData?: boolean;
  aggregationType?: 'sum' | 'mean' | 'median';
  timeframe?: 'daily' | 'weekly' | 'monthly';
} 

File: backend/app/schemas/report.py
from pydantic import BaseModel, Field, validator
from typing import Dict, Any, Optional, List, Union, Literal
from datetime import datetime

class StatisticalSummary(BaseModel):
    mean: Optional[float]
    median: Optional[float]
    std: Optional[float]
    min: Optional[float]
    max: Optional[float]
    count: Optional[int]
    percentiles: Optional[Dict[str, float]]

class TimeSeriesAnalysis(BaseModel):
    trend: str
    seasonality: str
    outliers: List[float]
    forecast: Optional[List[float]]
    trends: Optional[Dict[str, Dict[str, Any]]]

class NumericalAnalysis(BaseModel):
    mean: float
    median: float
    std: float
    min: float
    max: float
    quartiles: tuple[float, float, float]

class CategoricalAnalysis(BaseModel):
    unique_values: int
    most_common: List[tuple[str, int]]
    distribution: Dict[str, int]

class DataQuality(BaseModel):
    completeness: Dict[str, Any]
    accuracy: Dict[str, Any]
    consistency: Dict[str, Any]

class VisualizationData(BaseModel):
    type: str
    data: List[Dict[str, Any]]
    layout: Dict[str, Any]

class Analysis(BaseModel):
    llm_analysis: str
    insights: Dict[str, StatisticalSummary]
    statistical_analysis: Dict[str, Optional[Union[TimeSeriesAnalysis, NumericalAnalysis, CategoricalAnalysis]]]
    data_quality: DataQuality

class EnhancedReport(BaseModel):
    id: str
    data_type: Literal['time_series', 'numerical', 'categorical', 'mixed']
    analysis: Analysis
    visualizations: List[VisualizationData]
    metadata: Dict[str, Any]
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)

    @validator('data_type')
    def validate_data_type(cls, v):
        allowed_types = ['time_series', 'numerical', 'categorical', 'mixed']
        if v not in allowed_types:
            raise ValueError(f'data_type must be one of {allowed_types}')
        return v